File: fungal/src/core/self_model.cpp
#include "core/self_model.hpp"
#include <cmath>
#include <algorithm>

namespace fungal::core {

SelfModel::SelfModel() {
    // Initialize with one task type (v1: single type only)
    models_.resize(1, TaskTypeStats{});
}

double SelfModel::predict_success(int task_type_id) {
    auto& stats = get_or_create(task_type_id);
    // Return μ as the prediction
    return stats.success_mean;
}

double SelfModel::get_uncertainty(int task_type_id) const {
    if (task_type_id >= static_cast<int>(models_.size())) {
        return 0.3;  // default if not found
    }
    return models_[task_type_id].success_stddev;
}

void SelfModel::update_from_outcome(int task_type_id, bool outcome_correct, double predicted_prob) {
    auto& stats = get_or_create(task_type_id);

    // Update belief: Bayesian adjustment of μ and σ
    update_belief(stats, outcome_correct);

    // ACCURACY: did our strategy's claim match ground truth?
    // Separate signal: binary outcome of "was the claim right?"
    stats.total_predictions++;
    if (outcome_correct) {
        stats.accurate_predictions++;
    }

    // CALIBRATION: is our predicted probability justified by empirical frequency?
    // Track empirical success rate separately from accuracy
    double alpha = 0.15;  // smoothing factor for frequency tracking
    stats.empirical_success_rate = alpha * (outcome_correct ? 1.0 : 0.0) +
                                   (1.0 - alpha) * stats.empirical_success_rate;

    // Calibration error: how far is predicted probability from empirical frequency?
    // This measures: "when we predict P(success), do we actually see that frequency?"
    // Example: predict 70% but empirical is 40% → error = 0.3
    stats.calibration_error = std::abs(predicted_prob - stats.empirical_success_rate);
}

double SelfModel::get_calibration_error(int task_type_id) const {
    if (task_type_id >= static_cast<int>(models_.size())) {
        return 0.0;
    }
    return models_[task_type_id].calibration_error;
}

double SelfModel::get_accuracy(int task_type_id) const {
    if (task_type_id >= static_cast<int>(models_.size())) {
        return 0.5;
    }
    const auto& stats = models_[task_type_id];
    if (stats.total_predictions == 0) {
        return 0.5;
    }
    return static_cast<double>(stats.accurate_predictions) / stats.total_predictions;
}

double SelfModel::get_empirical_success_rate(int task_type_id) const {
    if (task_type_id >= static_cast<int>(models_.size())) {
        return 0.5;
    }
    return models_[task_type_id].empirical_success_rate;
}

TaskTypeStats& SelfModel::get_or_create(int task_type_id) {
    if (task_type_id < 0) {
        task_type_id = 0;
    }
    if (task_type_id >= static_cast<int>(models_.size())) {
        models_.resize(task_type_id + 1, TaskTypeStats{});
    }
    return models_[task_type_id];
}

TaskTypeStats SelfModel::get_stats(int task_type_id) const {
    if (task_type_id < 0 || task_type_id >= static_cast<int>(models_.size())) {
        return TaskTypeStats{};
    }
    return models_[task_type_id];
}

std::vector<TaskTypeStats> SelfModel::export_all() const {
    return models_;
}

void SelfModel::import_all(const std::vector<TaskTypeStats>& stats) {
    if (stats.empty()) {
        models_.assign(1, TaskTypeStats{});
        return;
    }
    models_ = stats;
}

void SelfModel::update_belief(TaskTypeStats& stats, bool outcome_correct) {
    // Bayesian update: outcome informs μ and σ
    // μ: mean success probability
    // σ: uncertainty in that estimate

    double outcome_value = outcome_correct ? 1.0 : 0.0;

    // Update mean: move toward observed outcome with adaptive learning rate
    // Higher uncertainty → faster learning (more plasticity when uncertain)
    double learning_rate = 0.15 * (1.0 + stats.success_stddev);
    stats.success_mean = (1.0 - learning_rate) * stats.success_mean +
                         learning_rate * outcome_value;

    // Update uncertainty based on prediction error
    // If prediction was close to outcome, reduce uncertainty (confidence increases)
    // If prediction was far, increase uncertainty (we need more info)
    double prediction_error = std::abs(outcome_value - stats.success_mean);

    // Scale σ adjustment by prediction error (not fixed multipliers)
    // Good prediction (error < 0.1): reduce σ by ~2%
    // Medium error (0.1-0.3): reduce σ by ~1%
    // Large error (>0.3): increase σ by error amount
    double sigma_adjustment;
    if (prediction_error < 0.1) {
        sigma_adjustment = 0.98;  // tighten confidence
    } else if (prediction_error < 0.3) {
        sigma_adjustment = 0.99;  // slight tightening
    } else {
        sigma_adjustment = 1.0 + (prediction_error * 0.2);  // scale up by error
    }
    stats.success_stddev *= sigma_adjustment;

    // Clamp σ to reasonable range [0.05, 0.5]
    stats.success_stddev = std::max(0.05, std::min(0.5, stats.success_stddev));
}

}  // namespace fungal::core

File: fungal_py/self_model.py
"""Bayesian self-model for success prediction and calibration."""

import math


class TaskTypeStats:
    """Statistics for a single task type."""
    def __init__(self):
        self.success_mean = 0.5  # μ: predicted success probability
        self.success_stddev = 0.2  # σ: uncertainty
        self.accurate_predictions = 0  # claim matched oracle?
        self.total_predictions = 0
        self.empirical_success_rate = 0.5
        self.calibration_error = 0.0


class SelfModel:
    """Bayesian predictor with separate accuracy and calibration tracking."""

    def __init__(self):
        self.task_stats = {}  # task_id -> TaskTypeStats

    def predict_success(self, task_id: int) -> float:
        """Predict probability of success for a task type."""
        if task_id not in self.task_stats:
            self.task_stats[task_id] = TaskTypeStats()
        return self.task_stats[task_id].success_mean

    def get_empirical_success_rate(self, task_id: int) -> float:
        """Return empirical success frequency over all cycles."""
        if task_id not in self.task_stats:
            return 0.5
        stats = self.task_stats[task_id]
        if stats.total_predictions == 0:
            return 0.5
        return stats.empirical_success_rate

    def get_calibration_error(self, task_id: int) -> float:
        """Return |predicted_prob - empirical_freq|."""
        if task_id not in self.task_stats:
            return 0.0
        return self.task_stats[task_id].calibration_error

    def update_from_outcome(self, task_id: int, outcome_correct: bool, predicted_prob: float):
        """Learn from observed outcome.

        Args:
            task_id: task type identifier
            outcome_correct: was the strategy claim correct? (binary)
            predicted_prob: what did self-model predict?
        """
        if task_id not in self.task_stats:
            self.task_stats[task_id] = TaskTypeStats()

        stats = self.task_stats[task_id]
        stats.total_predictions += 1

        # Accuracy: did claim match oracle?
        if outcome_correct:
            stats.accurate_predictions += 1

        # Empirical frequency tracking
        empirical_freq = stats.accurate_predictions / stats.total_predictions

        # Calibration: |predicted - empirical|
        stats.empirical_success_rate = empirical_freq
        stats.calibration_error = abs(predicted_prob - empirical_freq)

        # Update belief (μ/σ) with error-scaled Bayesian adjustment
        prediction_error = abs(predicted_prob - (1.0 if outcome_correct else 0.0))

        # Learning rate scales with uncertainty (σ)
        learning_rate = min(0.1, stats.success_stddev / 2.0)
        direction = 1.0 if outcome_correct else -1.0
        stats.success_mean += direction * learning_rate * (1.0 - stats.success_mean if direction > 0 else stats.success_mean)

        # Adjust σ: tight predictions that are correct → lower σ (98%)
        # Loose predictions or wrong predictions → higher σ
        if prediction_error < 0.1:
            stats.success_stddev *= 0.98
        elif prediction_error > 0.3:
            stats.success_stddev = min(0.5, stats.success_stddev * (1.0 + 0.2 * prediction_error))

        # Clamp to valid ranges
        stats.success_mean = max(0.0, min(1.0, stats.success_mean))
        stats.success_stddev = max(0.01, min(0.5, stats.success_stddev))
